service/src/main/java/io/camunda/service/UserTaskServices.java

```diff
@@ -40,11 +58,176 @@ public UserTaskServices withAuthentication(final Authentication authentication)
 
   @Override
   public SearchQueryResult<UserTaskEntity> search(final UserTaskQuery query) {
+    if (query.filter().variableFilters() != null && !query.filter().variableFilters().isEmpty()) {
+      final Map<Long, Set<Long>> scopeKeyMapByProcessInstanceKey = new HashMap<>();
+      final VariableServices variableServices =
+          new VariableServices(searchClient, transformers, authentication);
+      boolean firstIteration = true;
+
+      for (final VariableValueFilter variableValueFilter : query.filter().variableFilters()) {
+        processVariableFilter(
+            variableValueFilter, variableServices, scopeKeyMapByProcessInstanceKey, firstIteration);
+        firstIteration = false;
+      }
+
+      // Collect all unique processInstanceIds and scopeKeys
+      final List<Long> allProcessInstanceIds =
+          new ArrayList<>(scopeKeyMapByProcessInstanceKey.keySet());
+      final List<Long> allScopeKeys =
+          scopeKeyMapByProcessInstanceKey.values().stream()
+              .flatMap(set -> Optional.ofNullable(set).orElse(Collections.emptySet()).stream())
+              .filter(Objects::nonNull)
```

Comment:
:question: What happens here? Why is `filter(Objects::nonNull)` not sufficient?

Reply:
Okay, I think I got it. The `flatMap` ensures that a set is returned, and the `filter` filters values inside the set.